File: ios/Jazzify/Defense/Engine/DefenseSeparateTracksTransport.swift
import Foundation

enum DefenseSeparateTracksPhraseBars: Int {
    case one = 1
    case two = 2
    case four = 4
}

struct DefenseSeparateTracksGrid: Equatable {
    let cycleFrames: Int
    let cyclesPerForm: Int
    let bgmFrames: Int
    let sampleRate: Double
    let phraseBars: DefenseSeparateTracksPhraseBars
    let progressionBars: Int
    let playbackRatio: Double
}

struct DefensePhraseLoopWindow: Equatable {
    let rank: Int
    let startMeasure: Int
    let endMeasure: Int
    let sourceStartFrame: Int
    let sourceEndFrame: Int
}

struct DefensePhraseSchedule: Equatable {
    let targetCycle: Int
    let phraseIndex: Int
    let revision: Int
    let generation: UInt64
}

enum DefenseSeparateTracksTransport {
    static func idealCycleFrames(
        sampleRate: Double,
        bpm: Double,
        beatsPerBar: Int,
        phraseBars: Int,
        playbackRatio: Double
    ) -> Double {
        let safeFs = max(1.0, sampleRate)
        let safeBpm = max(1.0, bpm)
        let safeBeats = max(1, beatsPerBar)
        let safeK = max(1, phraseBars)
        let safeRatio = max(0.0001, playbackRatio)
        return safeFs * 60.0 / safeBpm * Double(safeBeats) * Double(safeK) / safeRatio
    }

    static func computeGrid(
        sampleRate: Double,
        bpm: Double,
        beatsPerBar: Int,
        phraseBars: DefenseSeparateTracksPhraseBars,
        progressionBars: Int,
        playbackRatio: Double
    ) -> DefenseSeparateTracksGrid {
        let cycleFrames = max(1, Int(idealCycleFrames(
            sampleRate: sampleRate,
            bpm: bpm,
            beatsPerBar: beatsPerBar,
            phraseBars: phraseBars.rawValue,
            playbackRatio: playbackRatio
        ).rounded()))
        let safeN = max(1, progressionBars)
        let cyclesPerForm = max(1, safeN / phraseBars.rawValue)
        return DefenseSeparateTracksGrid(
            cycleFrames: cycleFrames,
            cyclesPerForm: cyclesPerForm,
            bgmFrames: cyclesPerForm * cycleFrames,
            sampleRate: sampleRate,
            phraseBars: phraseBars,
            progressionBars: safeN,
            playbackRatio: playbackRatio
        )
    }

    static func measureSourceFrame(
        measureNumber: Int,
        bpm: Double,
        beatsPerBar: Int,
        sampleRate: Double
    ) -> Int {
        let safeMeasure = max(0, measureNumber)
        let safeBpm = max(1.0, bpm)
        let safeBeats = max(1, beatsPerBar)
        let safeFs = max(1.0, sampleRate)
        return Int(((Double(safeMeasure) * Double(safeBeats) * 60.0 / safeBpm) * safeFs).rounded())
    }

    static func isSourceFrameCountValid(actualFrames: Int, expectedFrames: Int) -> Bool {
        abs(actualFrames - expectedFrames) <= 1
    }

    static func phraseLoopWindow(
        rank: Int,
        phraseBars: DefenseSeparateTracksPhraseBars,
        bpm: Double,
        beatsPerBar: Int,
        sampleRate: Double
    ) -> DefensePhraseLoopWindow {
        let startMeasure = rank * phraseBars.rawValue + 1
        let endMeasure = (rank + 1) * phraseBars.rawValue
        return DefensePhraseLoopWindow(
            rank: rank,
            startMeasure: startMeasure,
            endMeasure: endMeasure,
            sourceStartFrame: measureSourceFrame(
                measureNumber: startMeasure - 1,
                bpm: bpm,
                beatsPerBar: beatsPerBar,
                sampleRate: sampleRate
            ),
            sourceEndFrame: measureSourceFrame(
                measureNumber: endMeasure,
                bpm: bpm,
                beatsPerBar: beatsPerBar,
                sampleRate: sampleRate
            )
        )
    }

    static func planPhraseReservation(
        absoluteCycle: Int,
        phaseFrame: Int,
        cycleFrames: Int,
        leadFrames: Int,
        phraseIndex: Int,
        revision: Int,
        generation: UInt64
    ) -> DefensePhraseSchedule {
        let safeF = max(1, cycleFrames)
        let safeLead = max(0, leadFrames)
        var targetCycle = absoluteCycle + 1
        var remaining = safeF - max(0, min(phaseFrame, safeF))
        while remaining < safeLead {
            targetCycle += 1
            remaining += safeF
        }
        return DefensePhraseSchedule(
            targetCycle: targetCycle,
            phraseIndex: phraseIndex,
            revision: revision,
            generation: generation
        )
    }

    static func beatInForm(
        absoluteCycle: Int,
        phaseFrame: Int,
        cycleFrames: Int,
        phraseBars: Int,
        beatsPerBar: Int,
        cyclesPerForm: Int
    ) -> Double {
        let safeF = max(1, cycleFrames)
        let safeQ = max(1, cyclesPerForm)
        let safeK = max(1, phraseBars)
        let safeBeats = max(1, beatsPerBar)
        let cycleInForm = ((absoluteCycle % safeQ) + safeQ) % safeQ
        let fraction = Double(max(0, min(phaseFrame, safeF))) / Double(safeF)
        return (Double(cycleInForm) * Double(safeK) + fraction * Double(safeK)) * Double(safeBeats)
    }

    static func bgmReadFrame(
        absoluteCycle: Int,
        phaseFrame: Int,
        cycleFrames: Int,
        cyclesPerForm: Int
    ) -> Int {
        let safeF = max(1, cycleFrames)
        let safeQ = max(1, cyclesPerForm)
        let cycleInForm = ((absoluteCycle % safeQ) + safeQ) % safeQ
        let safePhase = max(0, min(phaseFrame, safeF - 1))
        return cycleInForm * safeF + safePhase
    }
}

File: src/game/defense/defenseSeparateTracksTransport.ts
/**
 * Separate-tracks defense mode: K-cycle grid, phrase windows, reservation math.
 */

export type SeparateTracksPhraseBars = 1 | 2 | 4;

export interface SeparateTracksGrid {
  readonly cycleFrames: number;
  readonly cyclesPerForm: number;
  readonly bgmFrames: number;
  readonly beatFrames: number;
  readonly sampleRate: number;
  readonly phraseBars: SeparateTracksPhraseBars;
  readonly progressionBars: number;
  readonly playbackRatio: number;
}

export interface PhraseLoopWindow {
  readonly rank: number;
  readonly startMeasure: number;
  readonly endMeasure: number;
  readonly sourceStartFrame: number;
  readonly sourceEndFrame: number;
}

export interface PhraseSchedule {
  readonly targetCycle: number;
  readonly phraseIndex: number;
  readonly revision: number;
  readonly generation: number;
  readonly immediate: boolean;
}

export interface PlanPhraseReservationParams {
  readonly absoluteCycle: number;
  readonly phaseFrame: number;
  readonly cycleFrames: number;
  readonly beatFrames: number;
  readonly phraseIndex: number;
  readonly revision: number;
  readonly generation: number;
}

const PHRASE_BARS_VALUES: readonly SeparateTracksPhraseBars[] = [1, 2, 4];

export const isSeparateTracksPhraseBars = (value: number): value is SeparateTracksPhraseBars => (
  PHRASE_BARS_VALUES.includes(value as SeparateTracksPhraseBars)
);

export const computeIdealCycleFrames = (
  sampleRate: number,
  bpm: number,
  beatsPerBar: number,
  phraseBars: number,
  playbackRatio: number,
): number => {
  const safeFs = Math.max(1, sampleRate);
  const safeBpm = Math.max(1, bpm);
  const safeBeats = Math.max(1, beatsPerBar);
  const safeK = Math.max(1, phraseBars);
  const safeRatio = Math.max(0.0001, playbackRatio);
  return (safeFs * 60 / safeBpm * safeBeats * safeK) / safeRatio;
};

export const computeSeparateTracksGrid = (params: {
  readonly sampleRate: number;
  readonly bpm: number;
  readonly beatsPerBar: number;
  readonly phraseBars: SeparateTracksPhraseBars;
  readonly progressionBars: number;
  readonly playbackRatio: number;
}): SeparateTracksGrid => {
  const {
    sampleRate,
    bpm,
    beatsPerBar,
    phraseBars,
    progressionBars,
    playbackRatio,
  } = params;

  const cycleFrames = Math.round(computeIdealCycleFrames(
    sampleRate,
    bpm,
    beatsPerBar,
    phraseBars,
    playbackRatio,
  ));
  const safeF = Math.max(1, cycleFrames);
  const safeN = Math.max(1, progressionBars);
  const safeBeats = Math.max(1, beatsPerBar);
  const cyclesPerForm = Math.max(1, Math.trunc(safeN / phraseBars));
  const beatFrames = Math.max(1, Math.round(safeF / (safeBeats * phraseBars)));

  return {
    cycleFrames: safeF,
    cyclesPerForm,
    bgmFrames: cyclesPerForm * safeF,
    beatFrames,
    sampleRate,
    phraseBars,
    progressionBars: safeN,
    playbackRatio,
  };
};

export const computeMeasureSourceFrame = (
  measureNumber: number,
  bpm: number,
  beatsPerBar: number,
  sampleRate: number,
): number => {
  const safeMeasure = Math.max(0, measureNumber);
  const safeBpm = Math.max(1, bpm);
  const safeBeats = Math.max(1, beatsPerBar);
  const safeFs = Math.max(1, sampleRate);
  return Math.round((safeMeasure * safeBeats * 60 / safeBpm) * safeFs);
};

export const computePhraseLoopWindow = (
  rank: number,
  phraseBars: SeparateTracksPhraseBars,
  bpm: number,
  beatsPerBar: number,
  sampleRate: number,
): PhraseLoopWindow => {
  const startMeasure = rank * phraseBars + 1;
  const endMeasure = (rank + 1) * phraseBars;
  return {
    rank,
    startMeasure,
    endMeasure,
    sourceStartFrame: computeMeasureSourceFrame(startMeasure - 1, bpm, beatsPerBar, sampleRate),
    sourceEndFrame: computeMeasureSourceFrame(endMeasure, bpm, beatsPerBar, sampleRate),
  };
};

export const computeAllPhraseLoopWindows = (
  phraseCount: number,
  phraseBars: SeparateTracksPhraseBars,
  bpm: number,
  beatsPerBar: number,
  sampleRate: number,
): readonly PhraseLoopWindow[] => {
  const windows: PhraseLoopWindow[] = [];
  for (let rank = 0; rank < phraseCount; rank += 1) {
    windows.push(computePhraseLoopWindow(rank, phraseBars, bpm, beatsPerBar, sampleRate));
  }
  return windows;
};

export const computeExpectedSourceFrames = (
  barCount: number,
  bpm: number,
  beatsPerBar: number,
  sampleRate: number,
): number => (
  computeMeasureSourceFrame(barCount, bpm, beatsPerBar, sampleRate)
);

export const isSeparateTracksSourceFrameCountValid = (
  actualFrames: number,
  expectedFrames: number,
): boolean => Math.abs(actualFrames - expectedFrames) <= 1;

export interface PhraseLoopMeasureRow {
  readonly loopStartMeasure: number | null;
  readonly loopEndMeasure: number | null;
}

export const validatePhraseLoopMeasures = (
  phrasesSortedByOrderIndex: readonly PhraseLoopMeasureRow[],
  phraseBars: SeparateTracksPhraseBars,
): string | null => {
  if (phrasesSortedByOrderIndex.length === 0) {
    return 'separate tracks requires at least one phrase';
  }
  for (let rank = 0; rank < phrasesSortedByOrderIndex.length; rank += 1) {
    const expectedStart = rank * phraseBars + 1;
    const expectedEnd = (rank + 1) * phraseBars;
    const phrase = phrasesSortedByOrderIndex[rank];
    if (phrase.loopStartMeasure !== expectedStart || phrase.loopEndMeasure !== expectedEnd) {
      return `phrase rank ${rank} loop measures must be ${expectedStart}-${expectedEnd}`;
    }
  }
  return null;
};

export const validateSeparateTracksStageNumbers = (params: {
  readonly progressionBars: number | null;
  readonly phraseBars: number;
}): string | null => {
  const { progressionBars, phraseBars } = params;
  if (progressionBars === null || progressionBars <= 0) {
    return 'shared_progression_separate_tracks requires progressionBars';
  }
  if (!isSeparateTracksPhraseBars(phraseBars)) {
    return 'shared_progression_separate_tracks requires phraseBars in (1, 2, 4)';
  }
  if (progressionBars % phraseBars !== 0) {
    return 'progressionBars must be divisible by phraseBars';
  }
  return null;
};

export const planPhraseReservation = (
  params: PlanPhraseReservationParams,
): PhraseSchedule => {
  const {
    absoluteCycle,
    phaseFrame,
    cycleFrames,
    beatFrames,
    phraseIndex,
    revision,
    generation,
  } = params;

  const safeF = Math.max(1, cycleFrames);
  const safeBeat = Math.max(1, beatFrames);
  const safePhase = Math.max(0, Math.min(phaseFrame, safeF));

  if (safePhase <= safeBeat) {
    return {
      targetCycle: absoluteCycle,
      phraseIndex,
      revision,
      generation,
      immediate: true,
    };
  }

  const remainingFrames = safeF - safePhase;
  if (remainingFrames <= 0) {
    return {
      targetCycle: absoluteCycle,
      phraseIndex,
      revision,
      generation,
      immediate: true,
    };
  }

  return {
    targetCycle: absoluteCycle + 1,
    phraseIndex,
    revision,
    generation,
    immediate: false,
  };
};

export const computeBeatInForm = (
  absoluteCycle: number,
  phaseFrame: number,
  cycleFrames: number,
  phraseBars: number,
  beatsPerBar: number,
  cyclesPerForm: number,
): number => {
  const safeF = Math.max(1, cycleFrames);
  const safeQ = Math.max(1, cyclesPerForm);
  const safeK = Math.max(1, phraseBars);
  const safeBeats = Math.max(1, beatsPerBar);
  const cycleInForm = ((absoluteCycle % safeQ) + safeQ) % safeQ;
  const fraction = Math.max(0, Math.min(phaseFrame, safeF)) / safeF;
  return (cycleInForm * safeK + fraction * safeK) * safeBeats;
};

export const mapPhaseFrameForSpeedChange = (
  phaseFrame: number,
  oldCycleFrames: number,
  newCycleFrames: number,
): number => {
  const safeOld = Math.max(1, oldCycleFrames);
  const safeNew = Math.max(1, newCycleFrames);
  const fraction = Math.max(0, Math.min(phaseFrame, safeOld)) / safeOld;
  return Math.round(fraction * safeNew) % safeNew;
};

export const advanceTransportByElapsedFrames = (
  absoluteCycle: number,
  phaseFrame: number,
  cycleFrames: number,
  elapsedFrames: number,
): { readonly absoluteCycle: number; readonly phaseFrame: number } => {
  const safeF = Math.max(1, cycleFrames);
  const safePhase = Math.max(0, Math.min(phaseFrame, safeF));
  const safeElapsed = Math.max(0, elapsedFrames);
  const total = safePhase + safeElapsed;
  return {
    absoluteCycle: absoluteCycle + Math.floor(total / safeF),
    phaseFrame: total % safeF,
  };
};

export const computeBgmReadFrame = (
  absoluteCycle: number,
  phaseFrame: number,
  cycleFrames: number,
  cyclesPerForm: number,
): number => {
  const safeF = Math.max(1, cycleFrames);
  const safeQ = Math.max(1, cyclesPerForm);
  const cycleInForm = ((absoluteCycle % safeQ) + safeQ) % safeQ;
  const safePhase = Math.max(0, Math.min(phaseFrame, safeF - 1));
  return cycleInForm * safeF + safePhase;
};
